base/applications/mspaint/mouse.cpp

```diff
@@ -197,44 +240,89 @@ struct RectSelTool : ToolBase
     {
         if (bLeftButton)
         {
-            imageModel.ResetToPrevious();
             POINT pt = { x, y };
             imageModel.Bound(pt);
             selectionModel.SetRectFromPoints(start, pt);
-            RectSel(m_hdc, start.x, start.y, pt.x, pt.y);
+            imageModel.NotifyImageChanged();
         }
     }
 
     void OnButtonUp(BOOL bLeftButton, LONG x, LONG y) override
     {
         if (bLeftButton)
         {
-            imageModel.ResetToPrevious();
-            if (start.x == x && start.y == y)
-                imageModel.Undo(TRUE);
+            POINT pt = { x, y };
+            imageModel.Bound(pt);
+            selectionModel.SetRectFromPoints(start, pt);
             selectionModel.m_bShow = !selectionModel.m_rc.IsRectEmpty();
             imageModel.NotifyImageChanged();
         }
     }
 
     void OnFinishDraw() override
     {
-        m_bLeftButton = FALSE;
+        selectionModel.Landing();
         ToolBase::OnFinishDraw();
     }
 
     void OnCancelDraw() override
     {
-        if (m_bLeftButton)
-            imageModel.Undo(TRUE);
-        m_bLeftButton = FALSE;
+        selectionModel.HideSelection();
         ToolBase::OnCancelDraw();
     }
 };
 
-struct GenericDrawTool : ToolBase
+struct TwoPointDrawTool : ToolBase
 {
-    GenericDrawTool(TOOLTYPE type) : ToolBase(type)
+    BOOL m_bLeftButton = FALSE;
+    BOOL m_bDrawing = FALSE;
+
+    TwoPointDrawTool(TOOLTYPE type) : ToolBase(type)
+    {
+    }
+
+    void OnButtonDown(BOOL bLeftButton, LONG x, LONG y, BOOL bDoubleClick) override
+    {
+        m_bLeftButton = bLeftButton;
+        m_bDrawing = TRUE;
+        start.x = last.x = x;
+        start.y = last.y = y;
+        imageModel.NotifyImageChanged();
+    }
+
+    void OnMouseMove(BOOL bLeftButton, LONG x, LONG y) override
+    {
+        last.x = x;
```

Comment:
where is `last` declared? (and `start` in the function above) Is it a member of ToolBase? If so, why not using the usual `m_` prefix for these?

Reply:
They are global variables. Shall I add `g_` prefix?

Reply:
Added. Renamed some global variables.